codec-http2/src/main/java/io/netty/handler/codec/http2/ChildChannel.java

```diff
@@ -0,0 +1,252 @@
+        // Just remove the shutdownHook as this Channel may be closed later or registered to another
+        // EventLoop
+        ((SingleThreadEventLoop) eventLoop().unwrap()).removeShutdownHook(shutdownHook);
+    }
+
+    @Override
+    protected void doBeginRead() throws Exception {
+        if (readInProgress) {
+            return;
+        }
+
+        if (inboundBuffer.isEmpty()) {
+            readInProgress = true;
+            return;
+        }
+
+        final InternalThreadLocalMap threadLocals = InternalThreadLocalMap.get();
+        final Integer stackDepth = threadLocals.localChannelReaderStackDepth();
+        if (stackDepth < MAX_READER_STACK_DEPTH) {
+            threadLocals.setLocalChannelReaderStackDepth(stackDepth + 1);
+            try {
+                readTask.run();
+            } finally {
+                threadLocals.setLocalChannelReaderStackDepth(stackDepth);
+            }
+        } else {
+            eventLoop().execute(readTask);
+        }
+    }
+
+    @Override
+    protected final void doWrite(ChannelOutboundBuffer in) throws Exception {
+        if (closed) {
+            throw new ClosedChannelException();
+        }
+
+        EventLoop parentLoop = parent().eventLoop();
+
+        if (parentLoop.inEventLoop()) {
+            try {
+                for (;;) {
+                    Object msg = in.current();
+                    if (msg == null) {
+                        break;
+                    }
+                    parentWrite(msg);
+                    in.remove();
+                }
+                parentWriteComplete();
+            } catch (Throwable t) {
+                parentExceptionCaught(t);
+            }
+        } else {
+            // Use a copy because the original msgs will be recycled by AbstractChannel.
+            final Object[] msgsCopy = new Object[in.size()];
+            for (int i = 0; i < msgsCopy.length; i ++) {
+                msgsCopy[i] = ReferenceCountUtil.retain(in.current());
+                in.remove();
+            }
+
+            parentLoop.execute(new Runnable() {
+                @Override
+                public void run() {
+                    try {
+                        for (Object msg : msgsCopy) {
+                            parentWrite(msg);
+                            ReferenceCountUtil.release(msg);
+                        }
+                        parentWriteComplete();
+                    } catch (Throwable t) {
+                        parentExceptionCaught(t);
+                    }
+                }
+            });
+        }
+    }
+
+    /**
+     * Always called from the parent's event loop.
+     */
+    protected abstract void parentWrite(Object msg) throws Exception;
+
+    /**
+     * Always called from the parent's event loop.
+     */
+    protected abstract void parentWriteComplete() throws Exception;
+
+    protected abstract void parentExceptionCaught(Throwable t);
+
+    public void fireChildRead(Object msg) {
```

Comment:
Actually, since we are using the pipeline, the pipeline took care of running it on the proper place for us and so it didn't matter.

I'll try to reorganize a bit to make it clear which things need to run where.